python/node_classification.py

```diff
@@ -0,0 +1,744 @@
+    """
+    global model, current_values
+
+    # define fresh data
+    data = declare_data(ctx)
+
+    # define model
+    opt, criterion = declare_model(data)
+
+    current_values[TrainParams.NUM_EPOCHS] = no_epochs
+    num_nodes_sample = current_values[HeteroParams.NUM_NODES_SAMPLE]
+    num_iterations_sample = current_values[HeteroParams.NUM_ITERATIONS_SAMPLE]
+
+    # variables for early stopping
+    last_loss = 100
+    trigger_times = 0
+    last_time = time()
+    # training
+    for epoch in tqdm(range(1, no_epochs + 1)):
+        # one epoch of training, both training and validation loss are returned
+        loss, val_loss = train_epoch(
+            model,
+            opt,
+            data,
+            criterion,
+            current_values[TrainParams.BATCH_SIZE],
+            current_values[HeteroParams.OBSERVED_ATTRIBUTE],
+            {
+                key: [num_nodes_sample] * num_iterations_sample
+                for key in data.node_types
+            },
+        )
+
+        # early stopping
+        if val_loss > last_loss:
+            trigger_times += 1
+            print(
+                f"Loss has dropped for {trigger_times} time(s). Stopping after {current_values[OtherParams.PATIENCE] - trigger_times} more drops."
+            )
+
+            if trigger_times >= current_values[OtherParams.PATIENCE]:
+                print("Early stopping!")
+                break
+
+        else:
+            trigger_times = 0
+
+        last_loss = val_loss
+
+        global logged_data
+
+        # log data every console_log_freq epochs
+        if epoch % current_values[TrainParams.CONSOLE_LOG_FREQ] == 0:
+            model.eval()
+            out = model(data.x_dict, data.edge_index_dict)
+            dict_train = metrics(
+                data[current_values[HeteroParams.OBSERVED_ATTRIBUTE]].train_mask,
+                out,
+                data,
+                current_values[DataParams.METRICS],
+                current_values[HeteroParams.OBSERVED_ATTRIBUTE],
+                current_values[OtherParams.DEVICE_TYPE],
+            )
+            dict_val = metrics(
+                data[current_values[HeteroParams.OBSERVED_ATTRIBUTE]].val_mask,
+                out,
+                data,
+                current_values[DataParams.METRICS],
+                current_values[HeteroParams.OBSERVED_ATTRIBUTE],
+                current_values[OtherParams.DEVICE_TYPE],
+            )
+            logged_data.append(
+                {
+                    "epoch": epoch,
+                    "loss": loss,
+                    "val_loss": val_loss,
+                    "train": dict_train,
+                    "val": dict_val,
+                }
+            )
+
+            print(
+                f'Epoch: {epoch:03d}, Loss: {loss:.4f}, Val Loss: {val_loss:.4f}, Accuracy: {logged_data[-1]["train"]["accuracy"]:.4f}, Accuracy: {logged_data[-1]["val"]["accuracy"]:.4f}'
```

Comment:
Just make sure to run this file through black

Reply:
Yes, did it but it is somehow okay